### ISSUE TITLE
Listeners not removed

### ISSUE THREAD

**Issue by madbence** [othiym23/async-listener#37]
Honestly, i have absolutely no idea what's going on. (iojs 1.8.1)

A really simple `koa` app that demonstates the problem:

```js
'use strict';

require('async-listener');
const fs = require('fs');
const koa = require('koa');

let context = null;
let uid = 0;

const app = koa();

app
  .use(function* (next) {
    const ctx = this;
    const id = uid++;
    const listener = process.createAsyncListener({
      before: function () {
        fs.writeSync(1, 'enter ' + id + '\n');
        context = ctx;
      },
      after: function () {
        fs.writeSync(1, 'leave ' + id + '\n');
        context = null;
      },
    });
    console.log('- add listener %d', id);
    process.addAsyncListener(listener);
    context = ctx;
    yield* next;
    process.removeAsyncListener(listener);
    console.log('- removed listener %d', id);
  })
  .use(function* (next) {
    console.log(context ? context.url : 'no context');
    yield new Promise(function (resolve) {
      setTimeout(function () {
        resolve();
      });
    });
    console.log(context ? context.url : 'no context');
    yield* next;
  })
  .use(function* () {
    this.body = 'foobar';
  });

app.listen(3000);
```

The expected output (for an http req) would be (as far as i understand the concept of asynclisteners):

```
- add listener 0
/
enter 0
leave 0
...
enter 0
/
- removed listener 0
```

But the actual output shows, that for subsequent request, the previous callbacks are called too (i don't know why):

```
- add listener 1
/
enter 0
leave 0
enter 0
enter 1
leave 0
leave 1
enter 0
enter 1
leave 0
leave 1
enter 0
enter 1
/
- removed listener 1
leave 0
leave 1
enter 0
enter 1
leave 0
leave 1
enter 0
leave 0
enter 0
leave 0
enter 0
leave 0
enter 0
leave 0
```

If I put some `console.log` statements in `glue.js` after https://github.com/othiym23/async-listener/blob/master/glue.js#L481 It shows that the listeners are not removed.

It's *really* interesing, but if I put a `console.log` statement to https://github.com/othiym23/async-listener/blob/master/glue.js#L471, the listeners *are* removed correctly:

```js
if (!registered) listeners.push(listener);
console.log('after add, listeners: %d', listeners.length);
// require('fs').writeSync(1, 'after add, listeners: ' + listeners.length + '\n');
return listener;
```

Putting `console.log` after `process.addAsyncListener` call won't work. Using `fs.writeSync` instead of `console.log` also doesn't change the behavior.

Unfortunately, this `console.log` call breaks the expected asyncListener behavior:

```
after add, listeners: 1 // i guess this is logged because of promise inspection?
after remove, listeners: 0
- add listener 0 // console.log call in my example middleware
after add, listeners: 1 // console.log call in addAsyncListener in glue.js
/
enter 0
leave 0
enter 0
leave 0
enter 0
leave 0
no context // <-- this should print /
after remove, listeners: 0 // console.log call in removeAsyncListener in glue.js
- removed listener 0 // console.log call in my example middleware
enter 0
leave 0
- add listener 1
after add, listeners: 1
/
enter 1
leave 1
enter 1
leave 1
enter 1
leave 1
no context
after remove, listeners: 0
- removed listener 1
enter 1
leave 1
- add listener 2
after add, listeners: 1
/
enter 2
leave 2
enter 2
leave 2
enter 2
leave 2
no context
after remove, listeners: 0
- removed listener 2
enter 2
leave 2
```

Without `enter/leave`:

```
after add, listeners: 1
after remove, listeners: 0
- add listener 0
after add, listeners: 1
/
no context // <- this should print /
after remove, listeners: 0 // listeners are removed correctly
- removed listener 0
- add listener 1
after add, listeners: 1
/
no context
after remove, listeners: 0
- removed listener 1
- add listener 2
after add, listeners: 1
/
no context
after remove, listeners: 0
- removed listener 2
```

If I change the `console.log('after add, listeners: %d', listeners.length)` to `writeSync(1, 'after add, listeners: ' + listeners.length + '\n')`:

```
after add, listeners: 1
after remove, listeners: 0
- add listener 0
after add, listeners: 1
/
/ // <- output is fine
after remove, listeners: 0 // listener is removed?
- removed listener 0
- add listener 1
after add, listeners: 2 // the first listener is still there?
/
/
after remove, listeners: 1
- removed listener 1
- add listener 2
after add, listeners: 3
/
/
after remove, listeners: 2
- removed listener 2
```

 The `listeners` array keeps growing, up to 26 items.

Maybe I'm doing something wrong, but for me it seems that this is a bug :smile:

**Comment by hayes** [comment]
So this is a combination of a misunderstanding about how async-listener works with a known bug that should be fixed in #21, but I have more or less given up on getting that merged.  Here is a much simpler reproduction case, throwing koa and generators into the mix makes makes it easy to go down the wrong path while trying to debug this.

```javascript
'use strict';

require('async-listener');
var fs = require('fs');
var uid = 0;

setInterval(addOne, 1000);

function addOne() {
  var id = uid++;
  var listener = process.createAsyncListener({
    before: function () {
      fs.writeSync(1, 'enter ' + id + '\n');
    },
    after: function () {
      fs.writeSync(1, 'leave ' + id + '\n');
    },
  });

  process.addAsyncListener(listener);
  setTimeout(function noop() {
    process.removeAsyncListener(listener);
  });
}
```

One important thing to note about generators: they are NOT (guaranteed to be) synchronous.  This means that adding a listener, yielding, then removing the listener MAY be equivalent to removing the listener asynchronously.  Having an api (like generators) that is inconsistent about has no guarantees about its synchronicity is usually a recipe for headaches see [releasing zalgo](http://blog.izs.me/post/59142742143/designing-apis-for-asynchrony). 

In my simplified example above, the setTimeout after adding the listener is simulating the async yield.
removeListener is not documented as well as it should be, and the semantics of removing something asynchronously, are not obvious.  Basically, when a listener is added, that listener applies to all children of the current async context.  That means all callbacks passed to async functions in the current async context (tick, or whatever else you want to call it) will be wrapped.  These callbacks will inherit the same listener and apply it to any of its children.  Any callbacks that already existed, or are created from a context that does not have the listener will not be wrapped.  

Removing a listener works in the same way.  It does NOT apply globally, the listener is removed form the current context, and callbacks created from the current context will not be wrapped, any other context that have the same listener will continue to wrap callbacks.

for MOST cases you should never need to worry about removing a listener asynchronously, since it should only apply to calls of a specific chain, and removing it generally only affects a couple of callbacks.  The common case for remove is to add a listener, make a single async call, then remove the listener.  This means that that one callback will continue to propagate the listeners callbacks to its children, but nothing outside that call chain is affected. 

Back to the bug part of this issue:
async listener has some weird bugs around adding and removing listeners see test proving [add](https://github.com/othiym23/async-listener/pull/21/files#diff-c184271421b2029b538dbcd276550f99R1) and [remove](https://github.com/othiym23/async-listener/pull/21/files#diff-9948eb0aa3f8098d94b5e4d637f45c9dR39) have some bugs.

In this case, there is a bug that causes callbacks spawned while no listeners were added active (for example initial handler passed to the server created by koa) to all add to the same list of listeners.  There is a really simple, but hacky solution that should fix your issue.

add the following just after requiring `async-listener`
```javascript```
process.addAsyncListener({})
```

this will cause all callbacks to be wrapped as if there were already listeners, and behave as expected.
If you are interested in getting this fixed for real, go checkout #21, it needs to be updated to the latest changes, and fix a bit of logic for the 0.8 branch

Hope this is helpful

**Comment by madbence** [comment]
@hayes sir, you are awesome :trophy: 

You're absolutely right, listeners don't have to be removed explicitly, since they are *active* only in the current async context, which in this case means a single http call.
